Save about-the-study and files therein using generic result writer

flomnes opened the pull request.

flomnes committed:
Save about-the-study and files using generic result writer

flomnes committed:
Add getter for writer

flomnes committed:
Use generic writer to write problem.mps, criterion, interco, etc.

Those files are needed by Xpansion

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -773,51 +779,76 @@ bool Study::prepareOutput()
+  path.clear() << folderSettings << SEP << "generaldata.ini";
+  logs.notice() << path;
+  std::string generaldata_buffer;
+  Yuni::IO::File::LoadFromFile(generaldata_buffer, path);
```
What does a call to a **load** function do inside a **save** function ? Does it do a copy (as expected) ?

flomnes replied:
Yes, the idea is to dump the data into a file, then load the content into a buffer. I agree that this is not good, I'll change it.

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -773,51 +779,76 @@ bool Study::prepareOutput()
     return true;
 }
 
+void Study::saveMiscFilesIntoOutput()
```
I would rename this method into **save_about_the_study()**

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -145,6 +146,11 @@ void Study::clear()
     gotFatalError = false;
 }
 
+void Study::setWriter(Solver::IResultWriter::Ptr writer)
```
Why this function (**setWriter**) ?
I said in a previous review (#794 : src/solver/main/economy.cpp) that the writer should be created in the Application class, where it is first needed, and then passed to any component that needs it (simulation, xcast, ...)
Now, I would go a bit further : the writer should be created in the study, as soon as **study->parameters** gets the kind of writer should be buit, and then it should be passed to Application, simulation, and so on.

guilpier-code commented on src/libs/antares/study/study.h:
```diff
@@ -372,6 +372,8 @@ class Study final : public Yuni::NonCopyable<Study>, public IObject
     */
     bool prepareOutput();
 
+    void saveMiscFilesIntoOutput();
```
See **study.cpp** for remarks

guilpier-code commented on src/solver/application.cpp:
```diff
@@ -377,6 +377,10 @@ void Application::execute()
     }
 
+    // Save about-the-study files (comments, notes, etc.)
+    pStudy->setWriter(pResultWriter);
```
Please see comment in **study.cpp**

guilpier-code commented on src/libs/antares/study/simulation.cpp:
```diff
@@ -68,6 +68,14 @@ bool Simulation::saveToFolder(const AnyString& folder) const
     return false;
 }
 
+void Simulation::saveUsingWriter(Solver::IResultWriter::Ptr writer, const AnyString& folder) const
```
In method **Simulation::saveToFolder(...)**, instruction `b = folder;` is run twice.

guilpier-code commented on src/libs/antares/study/simulation.h:
```diff
@@ -64,6 +65,8 @@ class Simulation final
     */
     bool saveToFolder(const AnyString& folder) const;
 
+    void saveUsingWriter(Solver::IResultWriter::Ptr writer, const AnyString& folder) const;
```
Current class and source file have a terrible name (**Simulation**).
Actually, it's all about **user's notes**, loading and saving them (modifying them is handled in : 
src/ui/simulator/windows/notes.cpp).
It can be misleading. So this class should renamed **UserNotes**. Same for source file.
The content of user's notes should be renamed from **comments** into **content**.

guilpier-code replied:
Once "simulation" renamed, **saveUsingWriter** will be able to keep its name.

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -773,51 +779,76 @@ bool Study::prepareOutput()
+  path.reserve(1024);
+
+  path.clear() << "about-the-study";
+  simulation.saveUsingWriter(pResultWriter, path);
```
As said in **simulation.h**'s review, we should rename study's **simulation** data member into **user_notes_**

flomnes replied:
I've renamed it to `SimulationMetadata`

guilpier-code replied:
I'm sorry but **simulation meta data** doesn't speak to me. I still think **user's notes** is better. 
Question : how is it called in Antares Web ?
We should also change the name of the source file (simulation.h is unrelated to what this class does)

flomnes replied:
Renamed to `simulationComments`

guilpier-code commented on src/solver/application.cpp:
```diff
@@ -377,6 +377,10 @@ void Application::execute()
 
+    // Save about-the-study files (comments, notes, etc.)
+    pStudy->setWriter(pResultWriter);
+    pStudy->saveMiscFilesIntoOutput();
```
Reminder : 
in case we decide to initialize the writer in the study (and broadcast it in the rest of the program),  the call to **saveMiscFilesIntoOutput()** should be made (if **study.parameters** has the writer's info) where it was originally, that next to call to **prepareOutput()**

flomnes replied:
yes

flomnes committed:
Use generic writer for annual statistics (#808)

flomnes committed:
Some renaming

flomnes committed:
PR remarks

flomnes committed:
Finalize write upon receiving SIGINT / SIGTERM (#827)

* Bump to CXX 17

* Clarify execution-info.ini file

* Fix missing renewable columns in districts (sets of areas) (#802)

* execution_info.ini : add OR-Tools used & solver (#820)

* execution_info.ini : add OR-Tools used & solver

* PR remarks

* Fix wrong number of cores + code de-duplication (#793)

* Fix wrong number of cores + code de-duplication

* PR remarks + small refactor (template)

* Release/8.3.1 (#818)

* Finalize write upon receiving SIGINT / SIGTERM

This should ensure that a valid zip archive is produced even if the
user pushes "Ctrl+C".

* Avoid double-free

* Some formatting

* Refactor + add logs

* Simplify tuneSolverSpecificOptions (#829)

Use MPSolver::ProblemType

* [skip ci] Remove useless {}

* Restore notifications from Sonar (#828)

Source: https://github.com/AntaresSimulatorTeam/antares-xpansion/pull/501

* Statistics: use std::mutex / std::atomic<> to prevent concurrent writes (#838)

* Statistics: use std::mutex / std::atomic<> to prevent concurrent writes

* Apply PR remarks + renaming

* Base for BC dual values (#749)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* doc : new input and output format due to addition of BC marginal prices (#836)

* doc : new input and output format due to addition of BC marginal prices

* BC marginal prices doc : the folder containing the binding constraints output is renamed

* BC marginal price (#801)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

* Revert "BC marginal price (#801)"

This reverts commit a6fffd91c0087db4cb3d4532989f3adbb9e44047.

* Feature/bc marginal price (#845)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* BC. MARG. COST: use 2 decimals for the output value (#855)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* BC. MARG. COST: use 2 decimals for the output value

Reading the changelog of version 6.5, I came accros this fix note in
version 6.5

> - Output: the link-variable "MARG.COST" was rounded to an integer
  value (changed to 2 decimal accuracy)

So I thought that we might use 2 decimals for BC. MARG. COST as well.

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* Fix potential segfault with many BCs, refactor (#853)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* Fix potential segfault with many BCs, refactor

* Add static_cast<uint>

* Remove lambda return type

* Smell code

* Add TODO note

* PR remarks

* Use std::distance

* Code smells

* More code smells

* Remove unused member

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* Move Application::installSignalHandlers -> free function

* Add forgotten header file

* Use template helper function

* Fix build

Co-authored-by: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

flomnes committed:
Remove result writer setter/getter from Study

flomnes committed:
Fix link

flomnes committed:
Init writer in study

flomnes committed:
Fix overlapping zlib, 1st try

flomnes committed:
Revert "Fix overlapping zlib, 1st try"

This reverts commit ceec281cc70e93cf7b39d8f85d9046ac884e3171.

flomnes committed:
Use antares-deps 2.0.4-rc1

flomnes committed:
Use antares-deps 2.0.4-rc2

flomnes committed:
Add missing std:: (fix build)

flomnes committed:
Add forward decl for Benchmarking::IDurationCollector

flomnes committed:
Add & call Study::prepareWriter

flomnes committed:
Format code, fix build

flomnes committed:
Fix simple-study unit test

flomnes committed:
Use antares-deps 2.0.4

flomnes committed:
In-class init for pQueueService

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/study-writer

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/mps-generic-writer

flomnes committed:
Merge remote-tracking branch 'github/feature/mps-generic-writer' into fix/study-writer

flomnes committed:
minizip -> minizip-ng

flomnes committed:
Update antares-deps submodule

flomnes committed:
Revert multiple commits

flomnes committed:
Fix minizip

flomnes committed:
Revert to antares-deps 2.0.2

flomnes committed:
deps 2.0.3

flomnes committed:
Fix MZ flags

flomnes committed:
Force build of MZ in Windows

flomnes committed:
Fix configure command for Win

flomnes committed:
Use antares-deps 2.0.2

flomnes committed:
Use antares-deps 2.0.2 (no minizip)

flomnes committed:
Enable code coverage

flomnes committed:
Merge remote-tracking branch 'github/fix/overlapping-zlib-deps' into HEAD

flomnes committed:
Add try..catch to handle no queue/no result writer

flomnes committed:
Remove check (moved to init)

flomnes committed:
Merge branch 'feature/mps-generic-writer' into fix/study-writer

flomnes committed:
Save "about the study" earlier

flomnes reviewed (commented):
@guilpier-code DONE

flomnes committed:
Rename simulation -> simulationMetadata

flomnes commented:
@guilpier-code DONE

flomnes committed:
Use generic writer to write simulation.log (#805)

* Use generic writer to write simulation.log

* PR remarks

flomnes committed:
Merge remote-tracking branch 'github/feature/thread-writer-zip' into feature/about-the-study-generic-writer

flomnes committed:
MPS : don't attempt to create directory

flomnes committed:
Fix conficts

flomnes committed:
Merge remote-tracking branch 'github/feature/about-the-study-generic-writer' into feature/about-the-study-generic-writer

flomnes committed:
Use generic writer to write MPS files, criterion, interco, etc. (#807)

* Use generic writer for hydro debug files (#809)

* Use generic writer for hydro debug files

* PR remarks

* Remove result writer setter/getter from Study

* Fix link

* Init writer in study

* Add missing std:: (fix build)

* Add forward decl for Benchmarking::IDurationCollector

* Add & call Study::prepareWriter

* Format code, fix build

* Fix simple-study unit test

* In-class init for pQueueService

* PR remarks

* Add try..catch to handle no queue/no result writer

* Remove check (moved to init)

* Fix build

* Move writer creation to Study (#864)

* Remove result writer setter/getter from Study

* Fix link

* Init writer in study

* Add missing std:: (fix build)

* Add forward decl for Benchmarking::IDurationCollector

* Add & call Study::prepareWriter

* Format code, fix build

* Fix simple-study unit test

* In-class init for pQueueService

* Add try..catch to handle no queue/no result writer

* Remove check (moved to init)

* Remove unnecessary checks

flomnes committed:
Fix build

flomnes committed:
PR remarks

flomnes committed:
Rename saveMiscFilesIntoOutput -> saveAboutTheStudy

flomnes commented:
@guilpier-code DONE

flomnes committed:
Merge remote-tracking branch 'github/feature/about-the-study-generic-writer' into feature/about-the-study-generic-writer
